### PR TITLE
Improve product keys select

### PR DESCRIPTION
**Pull request by TsankoBg**
In this PR:
- load product keys with AJAX
- move `getProductKeys` to Woocommerce.php
- Add `LIMIT` to `getCustomAttributes` query to make it more efficient
- get rid of nested foreach for checking if a custom attribute is already in the list

[ch10719]

### CODE REVIEW

File: common/templates/options.php [hunk 1]
```diff
@@ -169,11 +169,9 @@
                         GTIN/EAN key
                         <select name="<?= $plugin->getOptionName('custom_gtin'); ?>">
```
Comment (apeschar) [inline common/templates/options.php]: * Instead of the `<SELECT>`, render a loading message until it's there:
  * So that the user knows it's loading.
  * So that if the form is submitted before the load completes, the value won't be reset.
* If the loading fails, show an error message instead of the select.

File: common/templates/options.php [hunk 2]
```diff
@@ -203,6 +201,23 @@
 </form>
 <script>
     (function ($) {
+        $.get("admin-ajax.php",
+            {
+                action: <?= json_encode($plugin->woocommerce->getProductKeysAction()); ?>,
+                selected_key: <?= json_encode($config["custom_gtin"]); ?>,
+            }
```
Comment (apeschar) [inline common/templates/options.php]: Encode the entire object with `json_encode`. It's shorter and avoids formatting accidents (missing commas, etc.).

File: common/src/WooCommerce.php [hunk 3]
```diff
@@ -464,6 +465,100 @@ public function getManualSyncNonce(): string {
...
+    public function getProductKeysAction(): string {
+        return $this->plugin->getOptionName('product_keys');
+    }
+
+    public function getProductKeys() {
+        $selected_key = $_GET['selected_key'];
+        wp_send_json([
+            'status' => true,
+            'data' => array_map(
+                function ($value) use ($selected_key) {
```
Comment (apeschar) [inline common/src/WooCommerce.php]: You can add a type hint on `$value` and on the return value here.

File: common/src/WooCommerce.php [hunk 4]
```diff
@@ -464,6 +465,100 @@ public function getManualSyncNonce(): string {
...
+            ON meta.post_id = posts.id 
+            WHERE posts.post_type = 'product' 
+            AND meta.meta_key='_product_attributes'
+            ORDER BY posts.id DESC
+            LIMIT 1000;
+        ";
+
+        $data = $wpdb->get_results($sql);
+        foreach ($data as $value) {
+            $product_attr = unserialize($value->meta_value);
```
Comment (apeschar) [inline common/src/WooCommerce.php]: To make this a little safer (unserialization can cause errors with certain values, non-existent classes, etc.), you can do this:

~~~php
try {
  $product_attr = unserialize($value->meta_value, ['allowed_classes' => false]);
} catch (\Throwable $e) {
  continue;
}
~~~

* Don't prevent unserialization of classes.
* If an error occurs during unserialization, just ignore it.

**Review by apeschar — CHANGES_REQUESTED** [review]

File: common/templates/options.php (hunk 1 above)
Comment (TsankoBg) [reply common/templates/options.php]: First I did that, but then I decided to disable and hide the `<select>` until load completes, instead of loading it entirely with jQuery

File: common/templates/options.php [hunk 5]
```diff
@@ -203,6 +205,33 @@
...
+                       options[property]["option_value"],
+                       options[property]["selected"],
+                       options[property]["selected"]
+                   );
+                   $(`.${options[property]["suggested"] ? "webwinkelkeur-suggested-keys" : "webwinkelkeur-other-keys"}`)
+                       .append(option);
+               }
+            }
+            $(".spinner").removeClass("is-active");
+            $('[name=<?= json_encode($plugin->getOptionName("custom_gtin")); ?>]')
```
Comment (apeschar) [inline common/templates/options.php]:
~~~
<?= json_encode(sprintf('[name="%s"]', $plugin->getOptionName("custom_gtin"))); ?>
~~~

File: common/templates/options.php [hunk 6]
```diff
@@ -203,6 +205,33 @@
 </form>
 <script>
     (function ($) {
+        $.get("admin-ajax.php",
+            <?= json_encode([
+                'action' => $plugin->woocommerce->getProductKeysAction(),
+                'selected_key' => $config["custom_gtin"],
+            ]); ?>,
```
Comment (apeschar) [inline common/templates/options.php]: Trailing comma?

File: common/templates/options.php [hunk 7]
```diff
@@ -203,6 +205,33 @@
 </form>
 <script>
     (function ($) {
+        $.get("admin-ajax.php",
```
Comment (apeschar) [inline common/templates/options.php]: Two options. Two. Not more:

~~~
FN(A, B, C)
~~~

~~~
FN(
  A,
  B,
  C
)
~~~

That's it.

File: common/templates/options.php [hunk 8]
```diff
@@ -166,17 +166,19 @@
                 <th scope="row"></th>
                 <td>
                     <label>
-                        GTIN/EAN key
-                        <select name="<?= $plugin->getOptionName('custom_gtin'); ?>">
-                            <option value=""><?= $plugin->getActiveGtinPlugin() ? __('Automatic detection', 'webwinkelkeur') . ' (' . (explode('/', $plugin->getActiveGtinPlugin())[0] ?? '') . ')' : 'Select key'; ?></option>
-                            <optgroup label="<?= _e('Suggested keys', 'webwinkelkeur'); ?>">
-                                <?= $plugin->getSelectOptions($config['custom_gtin'], true); ?>
+                        GTIN/EAN key:
+                        <select disabled hidden name=<?= json_encode($plugin->getOptionName('custom_gtin')); ?>>
```
Comment (apeschar) [inline common/templates/options.php]: WTF

~~~
name="<?= esc_attr($plugin->getOptionName('custom_gtin')); ?>"
~~~

File: common/templates/options.php [hunk 9]
```diff
@@ -203,6 +205,33 @@
...
+            ]); ?>,
+        ).done(function (response) {
+            if (!response || response.status === undefined) {
+                alert('Could not load product keys.');
+                return;
+            }
+            var options = response.data;
+            for (var property in options) {
+               if (options.hasOwnProperty(property)) {
+                   var option = new Option(options[property]["label"],
```
Comment (apeschar) [inline common/templates/options.php]:
~~~
.label
.option_value
.selected
~~~

**Review by apeschar — CHANGES_REQUESTED** [review]

**Review by apeschar — APPROVED** [review]
